// cohere-ai/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CohereTimeoutError = exports.CohereError = exports.CohereEnvironment = exports.Cohere = exports.SagemakerClient = exports.CohereClient = exports.BedrockClient = void 0;
var BedrockClient_1 = require("./BedrockClient");
Object.defineProperty(exports, "BedrockClient", { enumerable: true, get: function () { return BedrockClient_1.BedrockClient; } });
var Client_1 = require("./Client");
Object.defineProperty(exports, "CohereClient", { enumerable: true, get: function () { return Client_1.CohereClient; } });
var SagemakerClient_1 = require("./SagemakerClient");
Object.defineProperty(exports, "SagemakerClient", { enumerable: true, get: function () { return SagemakerClient_1.SagemakerClient; } });
exports.Cohere = __importStar(require("./api"));
var environments_1 = require("./environments");
Object.defineProperty(exports, "CohereEnvironment", { enumerable: true, get: function () { return environments_1.CohereEnvironment; } });
var errors_1 = require("./errors");
Object.defineProperty(exports, "CohereError", { enumerable: true, get: function () { return errors_1.CohereError; } });
Object.defineProperty(exports, "CohereTimeoutError", { enumerable: true, get: function () { return errors_1.CohereTimeoutError; } });
